unit/mr_test.c

```diff
@@ -48,58 +48,80 @@ static char err_buf[512];
  */
 static int mr_reg()
 {
-	int i;
+	int i, j;
 	int ret = 0;
-	int testret;
+	int testret = FAIL;
 	struct fid_mr *mr;
+	uint64_t access;
+	uint64_t *access_combinations;
 
-	testret = FAIL;
+	access = ft_info_to_mr_access(fi);
+	ft_alloc_bit_combo(0, access, &access_combinations);
 
 	for (i = 0; i < test_cnt; i++) {
 		buf_size = test_size[i].size;
-
-		// TODO test for various access
-		ret = fi_mr_reg(domain, buf, buf_size, ft_info_to_mr_access(fi), 0,
-				FT_MR_KEY, 0, &mr, NULL);
-		if (ret) {
-			FT_UNIT_STRERR(err_buf, "fi_mr_reg failed", ret);
-			goto fail;
-		}
-		ret = fi_close(&mr->fid);
-		if (ret) {
-			FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
-			goto fail;
+		for (j = 0; access_combinations[j] != 0; j++) {
+			ret = fi_mr_reg(domain, buf, buf_size, access_combinations[j], 0,
+					FT_MR_KEY, 0, &mr, NULL);
+			if (ret) {
+				FT_UNIT_STRERR(err_buf, "fi_mr_reg failed", ret);
+				goto fail;
+			}
+
+			ret = fi_close(&mr->fid);
+			if (ret) {
+				FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
+				goto fail;
+			}
 		}
 	}
 	testret = PASS;
 fail:
 	mr = NULL;
+	ft_free_bit_combo(access_combinations);
 	return TEST_RET_VAL(ret, testret);
 }
 
 static int mr_regv()
 {
-	int ret;
-	int testret;
+	int i, j, n;
+	int ret = 0;
+	int testret = FAIL;
 	struct fid_mr *mr;
-	struct iovec iov;
-
-	testret = FAIL;
-
-	iov.iov_base = buf;
-	iov.iov_len = test_size[test_cnt / 2].size;
+	struct iovec *iov;
+	char *base;
 
-	// TODO test for various iov counts
-	ret = fi_mr_regv(domain, &iov, 1, ft_info_to_mr_access(fi), 0,
-			FT_MR_KEY, 0, &mr, NULL);
-	if (ret) {
-		FT_UNIT_STRERR(err_buf, "fi_mr_regv failed", ret);
+	if (!fi->domain_attr->mr_iov_limit) {
+		ret = 1;
+		FT_UNIT_STRERR(err_buf, "fi_mr_regv failed, mr_iov_limit not defined", ret);
 		goto fail;
 	}
-	ret = fi_close(&mr->fid);
-	if (ret) {
-		FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
-		goto fail;
+
+	for (i = 0; i < test_cnt && test_size[i].size <= fi->domain_attr->mr_iov_limit; i++) {
+		n = test_size[i].size;
+		base = buf;
+		iov = malloc(sizeof(struct iovec) * n);
+		for (j = 0; j < n; j++) {
+			iov[j].iov_base = base;
+			iov[j].iov_len = test_size[test_cnt - 1].size / n;
+			base += iov[j].iov_len;			
+		}			
+
+		ret = fi_mr_regv(domain, &iov[0], n, ft_info_to_mr_access(fi), 0,
+				 FT_MR_KEY, 0, &mr, NULL);
+		if (ret) {
+			FT_UNIT_STRERR(err_buf, "fi_mr_regv failed", ret);
+			free(iov);
+			goto fail;
+		}
+
+		ret = fi_close(&mr->fid);
+		if (ret) {
+			free(iov);
+			FT_UNIT_STRERR(err_buf, "fi_close failed", ret);
+			goto fail;
+		}
+		free(iov);
```

Comment:
Move the free() call to the end, and remove duplicate calls above.  Use a single exit point to ensure that free is called.  (Or no more than 2 exit points, one for success and one for error).